src/services/services.ts

```diff
@@ -4459,6 +4497,466 @@ namespace ts {
 
+        function getImportModuleCompletionsAtPosition(fileName: string, position: number): ImportCompletionInfo {
+            synchronizeHostData();
+
+            const sourceFile = getValidSourceFile(fileName);
+            if (isInReferenceComment(sourceFile, position)) {
+                return getTripleSlashReferenceCompletion(sourceFile, position);
+            }
+            else if (isInString(sourceFile, position)) {
+                const node = findPrecedingToken(position, sourceFile);
+                if (!node || node.kind !== SyntaxKind.StringLiteral) {
+                    return undefined;
+                }
+
+                if (node.parent.kind === SyntaxKind.ImportDeclaration || isExpressionOfExternalModuleImportEqualsDeclaration(node) || isRequireCall(node.parent, false)) {
+                    // Get all known external module names or complete a path to a module
+                    return {
+                        entries: getStringLiteralCompletionEntriesFromModuleNames(<StringLiteral>node),
+                        textSpan: getDirectoryFragmentTextSpan((<StringLiteral>node).text, node.getStart() + 1)
+                    };
+                }
+            }
+
+            return undefined;
+
+            function getStringLiteralCompletionEntriesFromModuleNames(node: StringLiteral) {
+                const literalValue = normalizeSlashes(node.text);
+
+                const scriptPath = node.getSourceFile().path;
+                const scriptDirectory = getDirectoryPath(scriptPath);
+                if (isPathRelativeToScript(literalValue) || isRootedDiskPath(literalValue)) {
+                    const compilerOptions = program.getCompilerOptions();
+                    if (compilerOptions.rootDirs) {
+                        return getCompletionEntriesForDirectoryFragmentWithRootDirs(
+                            compilerOptions.rootDirs, literalValue, scriptDirectory, getSupportedExtensions(program.getCompilerOptions()), /*includeExtensions*/false, scriptPath);
+                    }
+                    else {
+                        return getCompletionEntriesForDirectoryFragment(
+                            literalValue, scriptDirectory, getSupportedExtensions(program.getCompilerOptions()), /*includeExtensions*/false, scriptPath);
+                    }
+                }
+                else {
+                    // Check for node modules
+                    return getCompletionEntriesForNonRelativeModules(literalValue, scriptDirectory);
+                }
+            }
+
+            /**
+             * Takes a script path and returns paths for all potential folders that could be merged with its
+             * containing folder via the "rootDirs" compiler option
+             */
+            function getBaseDirectoriesFromRootDirs(rootDirs: string[], basePath: string, scriptPath: string, ignoreCase: boolean): string[] {
+                // Make all paths absolute/normalized if they are not already

```

Comment:
I did some quick measurements to get an idea about whether this is too much IO. I generated a project with 200 node modules and 150 @ types modules and timed the completions for non-relative (e.g. "some-module-name") and relative (e.g. "../node-modules/") paths. In general, the time for completions was always sub 200 ms. For more reasonably sized directories, the completions are more or less instantaneous. There are rare spikes (up to about 600-700 ms) and I think those are likely due to accessing storage that has fallen out of memory or when the hard drive is in a cold state. Those are pretty rare, however, and after a spike happens you don't see another for a while. For the record, I was timing end-to-end (not just the IO).

Completions shouldn't be triggered on each keypress, just after trigger characters like "/". The caller should handle the filtering of the results until the next trigger character.